File: src/main/java/net/katsstuff/danmakucore/network/ChargeSpherePacket.java
/*
 * This class was created by <Katrix>. It's distributed as
 * part of the DanmakuCore Mod. Get the Source Code in github:
 * https://github.com/Katrix-/DanmakuCore
 *
 * DanmakuCore is Open Source and distributed under the
 * the DanmakuCore license: https://github.com/Katrix-/DanmakuCore/blob/master/LICENSE.md
 */
package net.katsstuff.danmakucore.network;

import io.netty.buffer.ByteBuf;
import net.katsstuff.danmakucore.helper.TouhouHelper;
import net.minecraft.client.Minecraft;
import net.minecraft.entity.Entity;
import net.minecraft.network.PacketBuffer;
import net.minecraftforge.fml.common.network.simpleimpl.IMessage;
import net.minecraftforge.fml.common.network.simpleimpl.IMessageHandler;
import net.minecraftforge.fml.common.network.simpleimpl.MessageContext;

public class ChargeSpherePacket {

	public static class Message implements IMessage {

		private int entity;
		private int amount;
		private double offset;
		private double divSpeed;
		private float r;
		private float g;
		private float b;
		private int lifetime;

		public Message(Entity entity, int amount, double offset, double divSpeed, float r, float g, float b, int lifetime) {
			this.entity = entity.getEntityId();
			this.amount = amount;
			this.offset = offset;
			this.divSpeed = divSpeed;
			this.r = r;
			this.g = g;
			this.b = b;
			this.lifetime = lifetime;
		}

		public Message() {}

		@Override
		public void fromBytes(ByteBuf byteBuf) {
			PacketBuffer buf = new PacketBuffer(byteBuf);
			entity = buf.readInt();
			amount = buf.readInt();
			offset = buf.readDouble();
			divSpeed = buf.readDouble();
			r = buf.readFloat();
			g = buf.readFloat();
			b = buf.readFloat();
			lifetime = buf.readInt();
		}

		@Override
		public void toBytes(ByteBuf byteBuf) {
			PacketBuffer buf = new PacketBuffer(byteBuf);
			buf.writeInt(entity);
			buf.writeInt(amount);
			buf.writeDouble(offset);
			buf.writeDouble(divSpeed);
			buf.writeFloat(r);
			buf.writeFloat(g);
			buf.writeFloat(b);
			buf.writeInt(lifetime);
		}
	}

	public static class Handler implements IMessageHandler<Message, IMessage> {

		@Override
		public IMessage onMessage(Message message, MessageContext ctx) {
			Minecraft.getMinecraft().addScheduledTask(() -> {
				Entity entity = Minecraft.getMinecraft().world.getEntityByID(message.entity);
				if(entity != null) {
					TouhouHelper.createChargeSphere(entity, message.amount, message.offset, message.divSpeed, message.r, message.g, message.b,
							message.lifetime);
				}
			});
			return null;
		}
	}
}

File: src/main/scala/net/katsstuff/teamnightclipse/danmakucore/network/ChargeSpherePacket.scala
/*
 * Copyright (C) 2018  Katrix
 * This file is part of DanmakuCore.
 *
 * DanmakuCore is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Lesser General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * DanmakuCore is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with DanmakuCore.  If not, see <https://www.gnu.org/licenses/>.
 */
package net.katsstuff.teamnightclipse.danmakucore.network

import net.katsstuff.teamnightclipse.mirror.network.scalachannel.{ClientMessageHandler, MessageConverter}
import net.katsstuff.teamnightclipse.danmakucore.scalastuff.TouhouHelper
import net.minecraft.client.Minecraft
import net.minecraft.client.network.NetHandlerPlayClient
import net.minecraft.entity.Entity

case class ChargeSpherePacket(
    entity: Int,
    amount: Int,
    offset: Double,
    divSpeed: Double,
    r: Float,
    g: Float,
    b: Float,
    lifetime: Int
) {

  def this(entity: Entity, amount: Int, offset: Double, divSpeed: Double, r: Float, g: Float, b: Float, lifetime: Int) =
    this(entity.getEntityId, amount, offset, divSpeed, r, g, b, lifetime)
}
object ChargeSpherePacket {

  implicit val converter: MessageConverter[ChargeSpherePacket] = MessageConverter.mkDeriver[ChargeSpherePacket].apply

  implicit val handler: ClientMessageHandler[ChargeSpherePacket, Unit] =
    new ClientMessageHandler[ChargeSpherePacket, Unit] {
      override def handle(netHandler: NetHandlerPlayClient, a: ChargeSpherePacket): Option[Unit] = {
        scheduler.addScheduledTask(ChargeSpherePacketRunnable(netHandler, a))
        None
      }
    }
}

case class ChargeSpherePacketRunnable(client: NetHandlerPlayClient, packet: ChargeSpherePacket) extends Runnable {
  override def run(): Unit = {
    val entity = Minecraft.getMinecraft.world.getEntityByID(packet.entity)
    if (entity != null)
      TouhouHelper
        .createChargeSphere(
          entity,
          packet.amount,
          packet.offset,
          packet.divSpeed,
          packet.r,
          packet.g,
          packet.b,
          packet.lifetime
        )
  }
}
